library/alloc/src/slice.rs

```diff
@@ -861,34 +861,37 @@ impl<T: Clone> ToOwned for [T] {
 // Sorting
 ////////////////////////////////////////////////////////////////////////////////
 
-/// Inserts `v[0]` into pre-sorted sequence `v[1..]` so that whole `v[..]` becomes sorted.
+/// Inserts `v[v.len() - 1]` into pre-sorted sequence `v[..v.len() - 1]` so that whole `v[..]` becomes sorted.
 ///
 /// This is the integral subroutine of insertion sort.
 #[cfg(not(no_global_oom_handling))]
-fn insert_head<T, F>(v: &mut [T], is_less: &mut F)
+// benchmarking indicated that inlining makes a substantial improvement, yet only requires a couple of hundred bytes
+#[inline(always)]
+fn insert_end<T, F>(v: &mut [T], is_less: &mut F)
 where
     F: FnMut(&T, &T) -> bool,
 {
-    if v.len() >= 2 && is_less(&v[1], &v[0]) {
+    let end = v.len() - 1;
```

Comment:
`saturating_sub` would prevent underflow.

Reply:
Good catch.  v.len() must be >= 2 for insert_end() to be called.  I will add a debug_assert:
```
    debug_assert!(v.len() >= 2);
```

Reply:
Do you think `saturating_sub` would degrade performance? Debug assertions are off generally, do I misunderstand how your fix prevents the underflow?

Reply:
If it did saturate that would be a bug.  Underflow might be better because it definitely would panic when indexing `v[end]`.  `insert_end` is never called unless `v.len() >= 2`, the debug_assert is there to indicate that, it is not possible for insert_end() to be called when `v.len() < 2`.

Reply:
Never called with those arguments, but that's not a precondition checked in this function (outside debug mode).
It was originally, on this line. If you perform fewer checks then that may improve performance, but people can recognize (part of) the performance gain is due to shortcuts taken rather than actual improvements. I'm not claiming you must not do that, but in general, I would err on the side of caution, especially around unsafe source code in the standard library. (There are other sorting libraries available.)

Reply:
I've added saturating_sub as requested, with no performance penalty.